ethcore/light/src/on_demand/mod.rs

```diff
@@ -227,20 +234,17 @@ impl Pending {
 		self.required_capabilities = capabilities;
 	}
 
-	// returning no reponse, it will result in an error.
-	// self is consumed on purpose.
-	fn no_response(self) {
-		trace!(target: "on_demand", "Dropping a pending query (no reply) at query #{}", self.query_id_history.len());
-		let err = self::error::ErrorKind::MaxAttemptReach(self.requests.num_answered());
+	// received too many empty responses, may be away to indicate a faulty request
+	fn bad_response(self, err: ResponseGuardError) {
+		let err = self::error::ErrorKind::BadResponse(err);
 		if self.sender.send(Err(err.into())).is_err() {
 			debug!(target: "on_demand", "Dropped oneshot channel receiver on no response");
```

Comment:
I could implement a logger for the circuit break that register every event:

```rust
pub trait Instrument {
    /// Calls when state machine reject a call.
    fn on_call_rejected(&self);

    /// Calls when the circuit breaker become to open state.
    fn on_open(&self);

    /// Calls when the circuit breaker become to half open state.
    fn on_half_open(&self);

    /// Calls when the circuit breaker become to closed state.
    fn on_closed(&self);
}
```

Sufficient?

Reply:
Otherwise, the error is actually propgated via RPC response but exactly how many bad responses and the particular reason is not propogated

Reply:
Having an 'Instrument' implementation writing in our logger looks like a good idea.

I also believe that being able to display the inner state of the circuit breaker (especially FailurePolicy debug info), is something that would greatly benefit the crate.

Reply:
Well, I found it useless because it doesn't get the internal state but I added better logs at least. 
Should be fine now, can you confirm? 

Example Request:
```bash
2018-10-31 12:12:38  jsonrpc-eventloop-0 TRACE on_demand  TransactionIndex Request, Incomplete { hash: Scalar(0xa17b45b011dad358740ccb65383e0900b086643d15192060f3983160492ccaff) }                                                           
2018-10-31 12:12:38  jsonrpc-eventloop-0 TRACE circuit_breaker  RequestGuard; failures: 1/10, state StateMachine { state: "closed" }                                                                                                          
2018-10-31 12:12:38  jsonrpc-eventloop-0 TRACE on_demand  Was unable to dispatch 1 requests.
2018-10-31 12:12:39  IO Worker #3 TRACE circuit_breaker  RequestGuard; failures: 2/10, state StateMachine { state: "closed" }                                                                                                                 
2018-10-31 12:12:39  IO Worker #3 TRACE on_demand  Was unable to dispatch 1 requests.
2018-10-31 12:12:44  IO Worker #0 TRACE circuit_breaker  RequestGuard; failures: 3/10, state StateMachine { state: "closed" }
2018-10-31 12:12:44  IO Worker #0 TRACE on_demand  Was unable to dispatch 1 requests.
2018-10-31 12:12:49  IO Worker #0 TRACE circuit_breaker  RequestGuard; failures: 4/10, state StateMachine { state: "open" }
2018-10-31 12:12:49  IO Worker #0 TRACE on_demand  Was unable to dispatch 1 requests.
2018-10-31 12:12:54  IO Worker #3 TRACE circuit_breaker  RequestGuard; failures: 4/10, state StateMachine { state: "open" }
2018-10-31 12:12:54  IO Worker #3 TRACE on_demand  Was unable to dispatch 1 requests.
2018-10-31 12:12:59  IO Worker #0 TRACE circuit_breaker  RequestGuard; failures: 5/10, state StateMachine { state: "open" }
....
2018-10-31 12:13:49  IO Worker #1 TRACE circuit_breaker  RequestGuard: reached_limit, failures: 10/10, state StateMachine { state: "open" }                                                                                                   
2018-10-31 12:13:49  IO Worker #1 TRACE on_demand  Was unable to dispatch 0 requests.
2018-10-31 12:13:49  jsonrpc-eventloop-0 DEBUG rpc  [Some(Num(1))] Took 70439ms
2018-10-31 12:13:49  jsonrpc-eventloop-0 DEBUG rpc  Response: {"jsonrpc":"2.0","error":{"code":-32065,"message":"Request time out"},"id":1}.   
```

Example Response:
```bash
2018-10-31 12:15:28  jsonrpc-eventloop-0 TRACE rpc  Request: {"method":"eth_getTransactionByHash","params":["0xa17b45b011dad358740ccb65383e0900b086643d15192060f3983160492ccaff"],"id":1,"jsonrpc":"2.0"}.
2018-10-31 12:15:28  jsonrpc-eventloop-0 TRACE on_demand  TransactionIndex Request, Incomplete { hash: Scalar(0xa17b45b011dad358740ccb65383e0900b086643d15192060f3983160492ccaff) }
2018-10-31 12:15:28  jsonrpc-eventloop-0 TRACE on_demand  Was unable to dispatch 0 requests.
2018-10-31 12:15:28  IO Worker #3 TRACE circuit_breaker  ResponseGuard: StateMachine { state: "closed" }, responses {EmptyResponse: 1}
2018-10-31 12:15:28  IO Worker #3 TRACE on_demand  Was unable to dispatch 0 requests.
2018-10-31 12:15:29  IO Worker #0 TRACE circuit_breaker  ResponseGuard: StateMachine { state: "closed" }, responses {EmptyResponse: 2}
2018-10-31 12:15:29  IO Worker #0 TRACE on_demand  Was unable to dispatch 0 requests.
2018-10-31 12:15:29  IO Worker #0 TRACE circuit_breaker  ResponseGuard: StateMachine { state: "closed" }, responses {EmptyResponse: 3}
...
2018-10-31 12:15:38  IO Worker #0 TRACE circuit_breaker  ResponseGuard: StateMachine { state: "open" }, responses: {EmptyResponse: 71}
2018-10-31 12:15:38  jsonrpc-eventloop-1 DEBUG rpc  [Some(Num(1))] Took 10310ms
2018-10-31 12:15:38  jsonrpc-eventloop-1 DEBUG rpc  Response: {"jsonrpc":"2.0","error":{"code":-32042,"message":"Bad response timeout, error cause was EmptyResponse, (majority count: 71 / total: 71)"},"id":1}.
```

Reply:
Nice to have some actual information about what happen :+1: